### DISCUSSION TITLE
How do we remove NFT’s from our AstroDAO wallet?

### DISCUSSION THREAD

**Discussion by NEARNFTDAO** [near-daos/astro-ui#655] [Q&A] [answered]
We would love to remove an NFT from our DAO NFT wallet but can’t seem to figure out how. Thank you.

**Comment by jaswinder6991** [comment]
Hey, you should be able to do this via custom function in Astro. By removing do you mean burn the NFT?
Can you share how you minted it? Mintbase or Paras?
Given that I can share a simple to-do for the burn.

**Reply by NEARNFTDAO** [reply]
What is a custom function for sending to another wallet

We put our Paras NFT’s in our DAO wallet, we would like to send 1 back to the original wallet for purpose of sale.

**Reply by jaswinder6991** [reply]
function name should be: **nft_transfer** if it is Paras.
You will have to put JSON arguments as:
{
  "token_id": "your token id",
  "receiver_id": "receiver's near name"
} 

Do you know your contract name?

**Reply by NEARNFTDAO** [reply]
No I do not. But the NFT’s in question are ones not minted on Paras but are traded there. Skellies for example. @jaswinder6991

**Reply by jaswinder6991** [reply]
Okay can you share the Paras link for the token? or any details you have?

**Reply by NEARNFTDAO** [reply]
https://paras.id/collection/secretskelliessociety.near

Here it is. Thank you again. @jaswinder6991

**Reply by jaswinder6991** [reply]
Cool. So you gotta do something like this on Astro custom function.
https://explorer.near.org/transactions/iVEFATUhx3HDLeS9cScHMVtTHYPkGi8j2TMCzicg6y6
What this means is you use the following details:
contract name: secretskelliessociety.near
function name: nft_transfer
JSON:
{
  "token_id": "72",  //put your token id here
  "receiver_id": "example.near". //put the near name of the receiver
}

**Reply by NEARNFTDAO** [reply]
![2CC4A0DB-E3E4-4A2F-8244-490F11099905](https://user-images.githubusercontent.com/99523884/154616134-77a375a5-c3cc-4a53-894d-06536ab6c2b1.jpeg)
Encountering this error, any suggestions?

**Reply by NEARNFTDAO** [reply]
@jaswinder6991

**Reply by NEARNFTDAO** [reply]
Hello! Any updates on the issue? 
@jaswinder6991

**Reply by NEARNFTDAO** [reply]
Any updates? Been 10 days, no response ser. If we uncovered a bug let us know. We have to inform out community. @jaswinder6991

**Reply by starpause** [reply]
Hmm reading https://nomicon.io/Standards/NonFungibleToken/ApprovalManagement I think that a step is missing before the `nft_transfer` function.

**Reply by NEARNFTDAO** [reply]
Thank you so much! Update us when possible. @starpause

**Comment by starpause** [comment, marked as answer]
Checking out the assets of the DAO in question, for reference:

https://app.astrodao.com/dao/near-nft-dao.sputnik-dao.near/treasury/nfts

### Gather info for Custom Function Call

It's not possible to get the `token_id` from the interface, which is a tracked feature enhancement that will be in verseion 3.0.3:

https://github.com/near-daos/astro-ui/issues/670

Until that feature ships, we can work around this by viewing the DAO treasury in Paras:

https://paras.id/near-nft-dao.sputnik-dao.near/collectibles

Let's say we want to transfer this NEAR Misfit:

https://paras.id/token/misfits.tenk.near::163

Let's remember `163` which is the `token_id`. The data we need to make a custom function call are the contract and it's parameters. Off to the explorer to see what transactions look like for the `misfits.tenk.near` contract:

https://explorer.near.org/accounts/misfits.tenk.near

What we're really interested in is a transfer 🔍

https://explorer.near.org/transactions/7TqiHddWiHqCUW7dD1vBUub7Y3ZBNyxNdxsbKeHgAV8U

This function call of `nft_transfer` has the parameters `token_id` and `receiver_id`:

![image](https://user-images.githubusercontent.com/408953/155921942-74b9c70e-12b4-469d-8cd9-4edf6838fff9.png)

Now we have:

Contract: `misfits.tenk.near`
Method/Function: `nft_transfer`
Variables: `token_id` = `163`, `receiver_id` = `boop.near`

Except you'd want to replace `boop.near` with a real account. 

### Make `nft_approve` Function call

Through trial and error we discovered that `nft_approve` is required before `nft_transfer` will work. This is documented in the nomicon:

https://nomicon.io/Standards/NonFungibleToken/ApprovalManagement

The most basic example provided is `nft::nft_approve({ "token_id": "1", "account_id": "bob" })`

I put a small deposit in the example because the nomicon example also has a small deposit in place:

![image](https://user-images.githubusercontent.com/408953/155925226-75e419a7-3a57-4cd8-84b1-f939ba8705f5.png)

Make a similar proposal, vote it through, then we're ready to move onto the `nft_transfer`. 

### Make `nft_transfer` Function call

In AstroDAO, hit the big green button to create a new proposal

<img width="582" alt="image" src="https://user-images.githubusercontent.com/408953/155922769-ef21ad20-72e9-4195-aa0c-1a05a7675175.png">

In "Propose a Transfer" use the "Proposal type" select scroll to the bottom and choose "Custom Function Call"

<img width="594" alt="image" src="https://user-images.githubusercontent.com/408953/155922916-32f0d888-7971-4bb7-bb48-fdb90fc07ebe.png">

Fill in all the info. Set TGas to 200 to avoid a "max gas exceeded" error, Set Deposit to "one yoctoNEAR" aka `0.000000000000000000000001` and hit "Propose"

![image](https://user-images.githubusercontent.com/408953/155936624-d2cfeb55-a23f-42c5-ad32-e8235b0e93e6.png)

Then have your council vote it through 😎

### Live Example

I followed these same steps with an NFT in the NEARchan DAO wallet:

`nft_approve` proposal:

https://app.astrodao.com/dao/nearchan.sputnik-dao.near/proposals/nearchan.sputnik-dao.near-14

`nft_transfer` proposal:

https://app.astrodao.com/dao/nearchan.sputnik-dao.near/proposals/nearchan.sputnik-dao.near-16

When we used the default of 150 TGas for the `nft_transfer` proposal we got an error that we didn't have enough gas. Hitting `Resubmit` did the trick and the transfer went through! I advise estimating gas on the low side (150 or 100) rather than trying to max it out, because when max gas is exceeded there is no "Resubmit" button. Your proposal will look passed in AstroDAO (since it did) but the call will fail and the transfer won't go through. 

![image](https://user-images.githubusercontent.com/408953/156034811-4ba0be88-13bc-4a18-94a0-39d92dac65ef.png)
